### PR TITLE
Filtering empty text examples

### PR DESCRIPTION
**Pull request by artemru**
**What does this PR do? Please describe:**
Removing empty text examples in ASR dataloaders
- This requires to move tokenization before bucketing!

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/datasets/asr.py [hunk 1]
```diff
@@ -160,6 +160,28 @@ def create_reader(
             self._name, split, builder.and_return(), gang, options
         )
 
+    @staticmethod
+    def add_tokenization_pipeline(
+        builder: DataPipelineBuilder,
+        tokenizer: TextTokenizer,
```
Comment (mattsetz) [inline src/fairseq2/datasets/asr.py:166]: we can add another arg here `remove_unk` to optionally remove all UNKs tokens from transcript. we should default this to False but it can be something that we experiment with.

File: src/fairseq2/datasets/asr.py [hunk 2]
```diff
@@ -160,6 +160,28 @@ def create_reader(
...
+
+        # to avoid to tokenize empty text, we filter out them out first
+        builder = builder.filter(lambda x: bool(len(x["text"]) > 0))
+
+        builder.map(text_encoder, selector="text")
+
+        unk_idx = tokenizer.vocab_info.unk_idx
+
+        def empty_text(example: Dict[str, Any]) -> bool:
+            return bool((example["text"] != unk_idx).sum().item() > 0)
```
Comment (mattsetz) [inline src/fairseq2/datasets/asr.py:180]: nit: `example["text"] != unk_idx).numel()` ?

File: src/fairseq2/datasets/asr.py (hunk 1 above)
Comment (mattsetz) [reply src/fairseq2/datasets/asr.py:166]: also, maybe we can log the number of UNKs (or % of `num_unks/total_tokens` within a sample/batch) in weights and biases? its useful for us to know if our dataset contains too many UNKs.

File: src/fairseq2/datasets/asr.py (hunk 2 above)
Comment (artemru) [reply src/fairseq2/datasets/asr.py:180]: no, it's not the same https://docs.pytorch.org/docs/stable/generated/torch.numel.html#torch-numel
example["text"] != unk_idx is boolean mask if we need to count True elements and all elements

File: src/fairseq2/datasets/asr.py (hunk 1 above)
Comment (artemru) [reply src/fairseq2/datasets/asr.py:166]: > also, maybe we can log the number of UNKs (or % of `num_unks/total_tokens` within a sample/batch) in weights and biases? its useful for us to know if our dataset contains too many UNKs.

it's possible (requires a bit to propagate some metric bags to data reader) but for now we can just do in a adhoc analysis

Comment (artemru) [reply src/fairseq2/datasets/asr.py:166]: > we can add another arg here `remove_unk` to optionally remove all UNKs tokens from transcript. we should default this to False but it can be something that we experiment with.

done !
